vitrivr-engine-index/src/main/kotlin/org/vitrivr/engine/index/decode/FFmpegVideoDecoder.kt

```diff
@@ -0,0 +1,230 @@
+                    source.metadata[Metadata.METADATA_KEY_AV_DURATION] = (videoStreamInfo.duration * 1000f).toLong()
+                    source.metadata[Metadata.METADATA_KEY_IMAGE_WIDTH] = videoStreamInfo.width
+                    source.metadata[Metadata.METADATA_KEY_IMAGE_HEIGHT] = videoStreamInfo.height
+                }
+
+                val audioStreamInfo = probeResult.streams.find { it.codecType == StreamType.AUDIO }
+
+                if (audioStreamInfo != null) {
+                    source.metadata[Metadata.METADATA_KEY_AUDIO_CHANNELS] = audioStreamInfo.channels
+                    source.metadata[Metadata.METADATA_KEY_AUDIO_SAMPLERATE] = audioStreamInfo.sampleRate
+                    source.metadata[Metadata.METADATA_KEY_AUDIO_SAMPLESIZE] = audioStreamInfo.sampleFmt
+                }
+
+                var windowEnd = TimeUnit.MILLISECONDS.toMicros(this@Instance.timeWindowMs)
+
+                val imageTransferBuffer = LinkedBlockingQueue<Pair<BufferedImage, Long>>(10)
+
+                val ffmpegInstance = ffmpeg.addInput(
+                    if (source is FileSource) {
+                        UrlInput.fromPath(source.path)
+                    } else {
+                        PipeInput.pumpFrom(source.newInputStream())
+                    }
+                ).addOutput(
+                    FrameOutput.withConsumerAlpha(
+                        object : FrameConsumer {
+
+                            val streamMap = mutableMapOf<Int, Stream>()
+
+                            override fun consumeStreams(streams: MutableList<Stream>) {
+                                streams.forEach { stream -> streamMap[stream.id] = stream }
+                            }
+
+                            override fun consume(frame: Frame) {
+
+                                val stream = streamMap[frame.streamId] ?: return
+
+                                when (stream.type) {
+                                    Stream.Type.VIDEO -> {
+                                        imageTransferBuffer.put(frame.image!! to (1000000 * frame.pts) / stream.timebase)
+                                    }
+
+                                    Stream.Type.AUDIO -> {
+                                        //TODO
+                                    }
+
+                                    null -> {
+                                        /* ignore */
+                                    }
+                                }
+                            }
+
+                        }
+                    )
+                ).setFilter(
+                    StreamType.VIDEO,
+                    "scale=w='min($maxWidth,iw)':h='min($maxHeight,ih)':force_original_aspect_ratio=decrease${if (framerate != null && framerate > 0) ",fps=$framerate" else ""}'"
+                )
+
+                //TODO audio settings
+
+                val future = ffmpegInstance.executeAsync()
```

Comment:
Not convinced that this the right way let alone the idiomatic way of doing this. 

Instead of creating and waiting for a future (within the co-routines `Flow`), wouldn't it be more sensible to let that `FrameConsumer` do the emission and to execute the FFmpeg process in a blocking fashion?

Reply:
This appears to not be possible since the external process is managed in a different thread that has no access to the flow context. I tried that and could not get it to work. It is entirely possible that I missed something, so if you manage to get it to work that way, feel free to change it accordingly.

Reply:
I think it is possible. I have made a proposal in my last commit.